irksome/imex.py

```diff
@@ -0,0 +1,267 @@
+                        repl[u0bits[k][ii]] = UUbits[j][k][ii]
+
+                # and sum the contribution
+                replF = replace(Ftmp, repl)
+                Fit += Ait[i, j] * dt * replF
+                Fprop += Aprop[i, j] * dt * replF
+    elif splitting == IA:
+        # diagonal contribution to iterator
+        for i in range(num_stages):
+            repl = {t: t+C[i]*dt}
+            for j in range(num_fields):
+                repl[u0bits[j]] = UUbits[i][j]
+                repl[vbits[j]] = VVbits[i][j]
+
+            # Also get replacements right for indexing.
+            for j in range(num_fields):
+                for ii in np.ndindex(u0bits[j].ufl_shape):
+                    repl[u0bits[j][ii]] = UUbits[i][j][ii]
+                    repl[vbits[j][ii]] = VVbits[i][j][ii]
+
+            Fit += dt * replace(Fexp, repl)
+
+        # dense contribution to propagator
+        Ablah = vc(np.linalg.solve(butch.A, Aexp))
+
+        for i in range(num_stages):
+            # replace test function
+            repl = {}
+
+            for k in range(num_fields):
+                repl[vbits[k]] = VVbits[i][k]
+                for ii in np.ndindex(vbits[k].ufl_shape):
+                    repl[vbits[k][ii]] = VVbits[i][k][ii]
+
+            Ftmp = replace(Fexp, repl)
+
+            # replace the solution with stage values
+            for j in range(num_stages):
+                repl = {t: t + C[j] * dt}
+
+                for k in range(num_fields):
+                    repl[u0bits[k]] = UUbits[j][k]
+                    for ii in np.ndindex(u0bits[k].ufl_shape):
+                        repl[u0bits[k][ii]] = UUbits[j][k][ii]
+
+                # and sum the contribution
+                Fprop += Ablah[i, j] * dt * replace(Ftmp, repl)
+    else:
+        raise NotImplementedError("Unsupported splitting type")
+
+    return Fit, Fprop
+
+
+class RadauIIAIMEXMethod():
+    """Class for advancing a time-dependent PDE via a polynomial
+    IMEX/RadauIIA method.  This requires one to split the PDE into
+    an implicit and explicit part.
+
+    The class sets up two methods -- `advance` and `iterate`.
+    The former is used to move the solution forward in time,
+    while the latter is used both to start the method (filling up
+    the initial stage values) and can be used at each time step
+    to increase the accuracy/stability.  In the limit as
+    the iterator is applied many times per time step,
+    one expects convergence to the solution that would have been
+    obtained from fully-implicit RadauIIA method.
+
+    :arg F: A :class:`ufl.Form` instance describing the implicit part
+            of the semi-discrete problem
+            F(t, u; v) == 0, where `u` is the unknown
+            :class:`firedrake.Function and `v` is the
+            :class:firedrake.TestFunction`.
+    :arg Fexp A :class:`ufl.Form` instance describing the part of the
+            PDE that is explicitly split off.
```

Comment:
I have no problem with separately specifying both implicit and explicit parts of the form, but I'm reminded that we had discussed using labels to specify this.  Was that something that was already done in Gusto that we were going to copy?

Reply:
Yes, there is form labeling in gusto.  It might be possible to use that as a fancy front-end from which the implicit and explicit parts.  If it works, it's just an interface/syntactic sugar change that can be done in a later PR.